pgrx/src/memcx.rs

```diff
@@ -45,3 +55,99 @@ where
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    #[cfg(not(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // Postgres 16 and newer permit any arbitrary power-of-2 alignment
+        unsafe {
+            // Bitflags for MemoryContextAllocAligned:
+            // #define MCXT_ALLOC_HUGE    0x01 /* allow huge allocation (> 1 GB) */
+            // #define MCXT_ALLOC_NO_OOM  0x02 /* no failure if out-of-memory */
+            // #define MCXT_ALLOC_ZERO    0x04 /* zero allocated memory */
+            let ptr = pgrx_pg_sys::MemoryContextAllocAligned(
+                self.ptr.as_ptr(),
+                layout.size(),
+                layout.align(),
+                0,
+            );
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, _layout: std::alloc::Layout) {
+        // TODO: Find faster free for use when MemoryContext is known.
+        // This is the global function that looks up the relevant Memory Context by address range.
+        pgrx_pg_sys::pfree(ptr.as_ptr().cast())
+    }
+
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    fn allocate_zeroed(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // Overriding default function here to use Postgres' zeroing implementation.
+        // On versions before Postgres 16, alignment is always 8 byte / 64 bit.
+        let layout = layout_for_pre16(layout)?;
+        unsafe {
+            let ptr = pgrx_pg_sys::MemoryContextAllocZero(self.ptr.as_ptr(), layout.size());
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    #[cfg(not(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")))]
+    fn allocate_zeroed(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // Overriding default function here to use Postgres' zeroing implementation.
+        // Postgres 16 and newer permit any arbitrary power-of-2 alignment
+        unsafe {
+            // Bitflags for MemoryContextAllocAligned:
+            // #define MCXT_ALLOC_HUGE    0x01 /* allow huge allocation (> 1 GB) */
+            // #define MCXT_ALLOC_NO_OOM  0x02 /* no failure if out-of-memory */
+            // #define MCXT_ALLOC_ZERO    0x04 /* zero allocated memory */
+            let ptr = pgrx_pg_sys::MemoryContextAllocAligned(
```

Comment:
Did you actually want to use `MemoryContextAllocZeroAligned` here?

Reply:
`MemoryContextAllocZeroAligned` is... *weird*. It existed before Postgres 16 introduced the arbitrary-alignment allocation API, and it does not take an alignment argument. The doc-string says 

> /*
> * MemoryContextAllocZeroAligned
> *      MemoryContextAllocZero where length is suitable for MemSetLoop
> *
> *  This might seem overly specialized, but it's not because newNode()
> *  is so often called with compile-time-constant sizes.
> */

[and it was removed from the Postgres 17 beta in a recent commit.](https://github.com/postgres/postgres/commit/3c080fb4fad3e1c1e34f74a7b84a443137adc9f2) 

I suspect it is misleadingly-named, and is instead "aligned" in the sense of "aligned for a very specific internal use-case"

Reply:
As this is, it's missing setting the allocated chunk to zeros, isn't it?

Reply:
> As this is, it's missing setting the allocated chunk to zeros, isn't it?

[Per Postgres documentation](https://doxygen.postgresql.org/palloc_8h.html), passing in bitflag 4 should zero the memory.

See: https://github.com/postgres/postgres/blob/master/src/backend/utils/mmgr/mcxt.c#L1256
https://github.com/postgres/postgres/blob/master/src/include/utils/palloc.h#L66